src/test/python/test_signer.py

```diff
@@ -0,0 +1,188 @@
+#   Copyright 2017 Covata Limited or its affiliates
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import pytest
+from covata.delta import signer
+import json
+import re
+from datetime import datetime
+import six.moves.urllib as urllib
+from covata.delta.signer import __encode_uri, SignatureMaterial
+from freezegun import freeze_time
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.asymmetric import padding
+from covata.delta import crypto
+
+freeze_time("2017-01-03 14:48:10")
+
+UNDESIRED_HEADERS = ["Connection", "Content-Length"]
+SIGNING_ALGORITHM = "CVT1-RSA4096-SHA256"
+CVT_DATE_FORMAT = "%Y%m%dT%H%M%SZ"
+
+
+def test_non_empty_payload():
+    assert signer.__get_hashed_payload(b'{"name": "rattan"}') == \
+    '4567de0d9e41eb8cc82ed6696700153adc78a43a4d09935d19e0804009d83737'
+
+
+def test_empty_payload():
+    assert signer.__get_hashed_payload(b'{}') == \
+    '44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a'
+
+
+def test_malformed_payload():
+    with pytest.raises(ValueError) as excepinfo:
+        signer.__get_hashed_payload(b'{""}')
+        assert excepinfo.type is json.JSONDecodeError
+
+
+def test_none_payload():
+    assert signer.__get_hashed_payload(b'{}') == \
+           signer.__get_hashed_payload(None)
+
+
+def test_unordered_payload():
+    assert signer.__get_hashed_payload(b'{"name": "rattan", "a": "value"}') == \
+    signer.__get_hashed_payload(b'{"a": "value", "name": "rattan"}')
+
+
+def test_nested_payload():
+    input_json = \
+        b"""
+        {
+            "name": "rattan",
+            "a": "value",
+            "c": {
+                "za": "value",
+                "a": "hello"
+            }
+        }
+        """
+
+    sorted_json = \
+        b"""
+        {
+            "a": "value",
+            "c": {
+                "a": "hello",
+                "za": "value"
+            },
+            "name": "rattan"
+        }
+        """
+
+    assert signer.__get_hashed_payload(input_json) == \
+           signer.__get_hashed_payload(sorted_json)
+
+
+def test_uri_encoding_with_space():
```

Comment:
as above, these 3 tests can be parameterized, i.e. `@pytest.mark.parameterized(...)`